contracts/P256Signer.sol

```diff
@@ -1,46 +1,81 @@
 pragma solidity ^0.8.0;
 
 import {Webauthn} from "./Webauthn.sol";
-import "hardhat/console.sol";
 
+/// @title P256Signer
+/// @notice A contract used to verify ECDSA signatures over secp256r1 through
+///         EIP-1271 of Webauthn payloads.
+/// @dev This contract is the implementation. It is meant to be used through
+///      proxy clone.
 contract P256Signer {
-  uint256 immutable public x;
-  uint256 immutable public y;
-
-  bytes4 constant internal EIP1271_MAGICVALUE = 0x1626ba7e;
-  bytes4 constant internal OLD_EIP1271_MAGICVALUE = 0x20c13b0b;
-
-  error InvalidSignature();
-  error InvalidHash();
-
-  constructor(uint256 _x, uint256 _y) {
-    x = _x;
-    y = _y;
-  }
-
-  function _validate(bytes memory data, bytes memory _signature) private view {
-    bytes32 _hash = keccak256(data);
-    (
-      bytes memory authenticatorData,
-      bytes memory clientData,
-      uint256 challengeOffset,
-      uint256[2] memory rs
-    ) = abi.decode(_signature, (bytes, bytes, uint256, uint256[2]));
-
-    bool valid = Webauthn.checkSignature(
-      authenticatorData, 0x01, clientData, _hash, challengeOffset, rs, [x, y]
-    );
-    
-    if (!valid) revert InvalidSignature();
-  }
-
-  function isValidSignature(bytes32 _hash, bytes memory _signature) public view returns (bytes4) {
-    _validate(abi.encode(_hash), _signature);
-    return EIP1271_MAGICVALUE;
-  }
-
-  function isValidSignature(bytes memory _hash, bytes memory _signature) public view returns (bytes4) {
-    _validate(_hash, _signature);
-    return OLD_EIP1271_MAGICVALUE;
-  }
+    /// @notice The EIP-1271 magic value
+    bytes4 internal constant EIP1271_MAGICVALUE = 0x1626ba7e;
+
+    /// @notice The old EIP-1271 magic value
+    bytes4 internal constant OLD_EIP1271_MAGICVALUE = 0x20c13b0b;
+
+    /// @notice Whether the contract has been initialized
+    bool public initialized;
+
+    /// @notice The x coordinate of the secp256r1 public key
+    uint256 public x;
+
+    /// @notice The y coordinate of the secp256r1 public key
+    uint256 public y;
+
+    /// @notice Error message when the signature is invalid
+    error InvalidSignature();
+
+    /// @notice Error message when the hash is invalid
+    error InvalidHash();
+
+    /// @notice Error message when the contract is already initialized
+    error AlreadyInitialized();
+
+    constructor() {
+        initialized = true;
```

Comment:
btw, it is not clear what the behavior of P256Signer is when x & y are 0. We might want to check

Reply:
From my understanding: 0,0 is not a valid point on secp256r1
Hence all signatures will be invalid

There is a check in FCL lib to ensure Q is on the curve 
* https://github.com/alembic-tech/p256-signer/blob/main/contracts/FCL/FCL_elliptic.sol#L700
* https://github.com/alembic-tech/p256-signer/blob/main/contracts/FCL/FCL_elliptic.sol#L228

It will reject if x or y == 0
```solidity
function ecAff_isOnCurve(uint x, uint y) internal pure returns (bool) {
    if (0 == x || x == p || 0 == y || y == p) {
        return false;
    }
   ...
}
```

Sage script verification

```
# Sage math secp256r1
a = Integer("0xFFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFC")
b = Integer("0x5AC635D8AA3A93E7B3EBBD55769886BC651D06B0CC53B0F63BCE3C3E27D2604B")
p = Integer("0xFFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFF")

K = GF(p)
E = EllipticCurve(K,[a,b])

x = Integer("0x0")
y = Integer("0x0")

P = E([x,y]) ### try to set the point with the coordinates.
```